crates/goose/src/agents/agent.rs

```diff
@@ -1,76 +1,804 @@
+            tool_result_tx: tool_tx,
+            tool_result_rx: Arc::new(Mutex::new(tool_rx)),
+        }
+    }
+
+    /// Get a reference count clone to the provider
+    pub fn provider(&self) -> Arc<dyn Provider> {
+        Arc::clone(&self.provider)
+    }
+
+    /// Check if a tool is a frontend tool
+    pub fn is_frontend_tool(&self, name: &str) -> bool {
+        self.frontend_tools.contains_key(name)
+    }
+
+    /// Get a reference to a frontend tool
+    pub fn get_frontend_tool(&self, name: &str) -> Option<&FrontendTool> {
+        self.frontend_tools.get(name)
+    }
+
+    /// Get all tools from all clients with proper prefixing
+    pub async fn get_prefixed_tools(&mut self) -> ExtensionResult<Vec<Tool>> {
+        let mut tools = self
+            .extension_manager
+            .lock()
+            .await
+            .get_prefixed_tools()
+            .await?;
+
+        // Add frontend tools directly - they don't need prefixing since they're already uniquely named
+        for frontend_tool in self.frontend_tools.values() {
+            tools.push(frontend_tool.tool.clone());
+        }
+
+        Ok(tools)
+    }
+
+    /// Dispatch a single tool call to the appropriate client
+    #[instrument(skip(tool_call, extension_manager, request_id), fields(input, output))]
+    async fn create_tool_future(
+        extension_manager: &ExtensionManager,
+        tool_call: mcp_core::tool::ToolCall,
+        is_frontend_tool: bool,
+        request_id: String,
+    ) -> (String, Result<Vec<Content>, ToolError>) {
+        let result = if tool_call.name == PLATFORM_READ_RESOURCE_TOOL_NAME {
+            // Check if the tool is read_resource and handle it separately
+            extension_manager
+                .read_resource(tool_call.arguments.clone())
+                .await
+        } else if tool_call.name == PLATFORM_LIST_RESOURCES_TOOL_NAME {
+            extension_manager
+                .list_resources(tool_call.arguments.clone())
+                .await
+        } else if tool_call.name == PLATFORM_SEARCH_AVAILABLE_EXTENSIONS_TOOL_NAME {
+            extension_manager.search_available_extensions().await
+        } else if is_frontend_tool {
+            // For frontend tools, return an error indicating we need frontend execution
+            Err(ToolError::ExecutionError(
+                "Frontend tool execution required".to_string(),
+            ))
+        } else {
+            extension_manager
+                .dispatch_tool_call(tool_call.clone())
+                .await
+        };
+
+        debug!(
+            "input" = serde_json::to_string(&tool_call).unwrap(),
+            "output" = serde_json::to_string(&result).unwrap(),
+        );
+
+        (request_id, result)
+    }
+
+    /// Truncates the messages to fit within the model's context window
+    /// Ensures the last message is a user message and removes tool call-response pairs
+    async fn truncate_messages(
```

Comment:
how are you thinking about other context management strategies here? If we included summarization or something, how does that fit in

Reply:
i wanna add in an agentconfig struct where users can specify context mgmt strategy, other params like max_truncation_attempts (this will be after refactoring reply)